### ISSUE TITLE
Replace C implementation of SHA-512 (and SHA-384) with a fast Rust implementation

### ISSUE THREAD

**Issue by briansmith** [briansmith/ring#62]
The minimum requirement is that the Rust implementation is approximately the same speed as the BoringSSL C (OPENSSL_NO_ASM) implementation and that the code is clear and that the licensing is clear.

I noticed that the rust-crypto implementation of SHA-2 (SHA-256, SHA-384, and SHA-512) is actually faster than the BoringSSL C implementation. But, it is using the MIT/Apache dual license. I contacted the author of the optimized version about relicensing as ISC via email, but I never heard back. Maybe there are other alternatives.

Note that we only need the block function, and SHA-384 and SHA-512 use the same block function.

**Comment by briansmith** [comment]
Also note that we already have an extensive test suite, so no new tests are needed.

**Comment by briansmith** [comment]
It might be better to leave the C implementations of the transform functions and instead use formal methods to prove the C bits correct.

**Comment by SimonSapin** [comment]
> The minimum requirement is that the Rust implementation is approximately the same speed as the BoringSSL C (OPENSSL_NO_ASM) implementation

In order to do this, how can one make sure `OPENSSL_NO_ASM` is enabled? Should it be exposed as a [Cargo feature](http://doc.crates.io/manifest.html#the-features-section)? (Or rather, since Cargo features are supposed to be additive, have an `asm` default feature and use `OPENSSL_NO_ASM` when `asm` is *not* enabled.)

**Comment by briansmith** [comment]
This isn't the kind of thing that should be controlled by a cargo feature, except maybe temporarily on Windows since Windows has toolchain issues with the asm code.

The main problem with doing this work is that it makes little practical difference since the Rust code would almost never be run, as every major platform (32-bit ARM, AAarch64, x86, x86-64) has assembly language code for this. Also, OpenSSL upstream has asm code for more platforms. So, what's the point? How can we really evaluate the performance? Perhaps we should just code these like the SHA-1 implementation--prefer clarity over performance. Then, maybe we can experiment with using some interesting technique like using macros to unroll the code without changing its overall form (i.e. without a bunch of copy-pasta that manual unrolling entails). Or, maybe we should try to use the Rust code as a case study for the Rust compiler's optimizer? I think any of these options are reasonable.

My perspective on this is that, technically, there's not much benefit to doing this, but politically it makes people feel good about *ring* because it makes *ring* seem less like a Rust wrapper around C code and more like a Rust library that has assembly language optimizations.

In other words, I think it's probably worth optimizing for clarity first, speed second, and maybe even defer the speed optimizations for later.

**Comment by briansmith** [comment]
Note that on Windows we don't currently use the asm code, but I'm guessing you don't develop on Windows. I guess implementing a no-asm feature, temporarily, wouldn't be the end of the world. It would probably get removed later. We can always just avoid documenting the feature.

**Comment by SimonSapin** [comment]
Alright. If you’d rather not expose `OPENSSL_NO_ASM` I can tweak makefiles locally to benchmark it.

**Comment by briansmith** [comment]
A lot of work on this has already been done. Note that the C implementation is long-gone.

**Comment by briansmith** [comment]
We already removed the C implementations, so this isn't an issue any more. The implementation in #199 will get merged (perhaps with changes) once we have a clear plan about what to do for Rust implementations of the loweset-level primitives.
